[TM] Port TG Jump Point Search and SSpathfinder

Vi3trice opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Ports https://github.com/tgstation/tgstation/pull/56780, and its associated subsystem.

Due to the nature of JPS, it has to be a uniform cost grid. This means diagonals are the same as horizontals, and due to the way the scanning works, bots will favor diagonals if there's a blocker nearby. Hence here, they will take a weird diagonal as the wall next to it gets detected, and the next closest valid tile is on a diagonal. The attempt at implementing cost didn't work quite properly for long distance, and thus broke mules.

If we're merging JPS, we'd have to accept that the bots will from then onward be a little quirky. Moreso with the 2-steps patrol routine that doesn't quite like diagonals.

![Bot](https://cdn.discordapp.com/attachments/145700319819464704/1016766365669011576/unknown.png)

The pathfinding overlay for bots doesn't expect diagonals, so it's something that would need fixing later.
![JPS_BEEP](https://user-images.githubusercontent.com/80771500/188756734-60274e46-3b4b-4d19-8254-1651c471356c.PNG)

## Why It's Good For The Game
A test run would be needed to see the actual perf cost.
Fixes #19322 as MULEs weren't actually incrementing their path when moving.

## Testing
Added obstacles all over, let the bots do their things. Summoning them as well since it forces them to deviate from their typical paths. Ran around lavaland as well to catch anything.

## Changelog
:cl:
tweak: Mob pathing can take diagonal directions, and should be faster for summoned bots.
fix: MULE pathing clears properly when they move around
fix: ED209s handle railing near the end of their paths
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

lewcc commented on code/__HELPERS/path.dm:
```diff
@@ -0,0 +1,433 @@
+		//	if(destination_turf.density)
+		//		return TRUE
+		if(TURF_PATHING_PASS_PROC)
+			if(!destination_turf.CanAStarPass(ID, actual_dir , caller, no_id = no_id))
```
```suggestion
			if(!destination_turf.CanAStarPass(ID, actual_dir, caller, no_id = no_id))
```

lewcc commented on code/__HELPERS/path.dm:
```diff
@@ -0,0 +1,433 @@
+	/// These are generally cheaper than looping contents so they go first
+	switch(destination_turf.pathing_pass_method)
+		// This is already assumed to be true
+		//if(TURF_PATHING_PASS_DENSITY)
+		//	if(destination_turf.density)
+		//		return TRUE
```
Commented out code here, intended?

lewcc commented on code/__HELPERS/path.dm:
```diff
@@ -0,0 +1,433 @@
+		var/first_step_direction_a = in_dir & 3 // eg. north   (1+8)&3 (0000 0011) = 1 (0000 0001)
+		var/first_step_direction_b = in_dir & 12 // eg. west   (1+8)&12 (0000 1100) = 8 (0000 1000)
+
+		for(var/first_step_direction in list(first_step_direction_a,first_step_direction_b))
+			var/turf/midstep_turf = get_step(destination_turf,first_step_direction)
+			var/way_blocked = midstep_turf.density || LinkBlockedWithAccess(midstep_turf,caller,ID, no_id = no_id) || midstep_turf.LinkBlockedWithAccess(destination_turf,caller,ID, no_id = no_id)
```
```suggestion
		for(var/first_step_direction in list(first_step_direction_a, first_step_direction_b))
			var/turf/midstep_turf = get_step(destination_turf, first_step_direction)
			var/way_blocked = midstep_turf.density || LinkBlockedWithAccess(midstep_turf, caller, ID, no_id = no_id) || midstep_turf.LinkBlockedWithAccess(destination_turf, caller, ID, no_id = no_id)
```

lewcc commented on code/game/turfs/turf.dm:
```diff
@@ -578,3 +582,24 @@
+	. = list()
+
+	for(var/iter_dir in GLOB.cardinal)
+		var/turf/turf_to_check = get_step(src,iter_dir)
```
```suggestion
		var/turf/turf_to_check = get_step(src, iter_dir)
```

lewcc commented on code/modules/mob/living/simple_animal/bot/bot.dm:
```diff
@@ -602,50 +603,53 @@ Pass a positive integer as an argument to override a bot's default speed.
-	else if(path.len > 0 && patrol_target)		// valid path
-		var/turf/next = path[1]
-		if(next == loc)
+	else if(path.len > 0 && patrol_target) // valid path
```
```suggestion
	else if(length(path) && patrol_target) // valid path
```

lewcc reviewed (commented):
Some nits, curious to see how much more performant it'll be over a*

Vi3trice commented:
Messing around, I found out that the bots passing through windows wasn't an issue with pathfinding, but one of how movement is handled by them.

Once a path is made and chosen by AStar/JPS, it's transfered over to the object and they'll handle it from there.

Vi3trice commented:
I added a fix for cleanbots as they were busted to begin with.

During the testmerge, it seems medbots had their paths nulled at some point, which shouldn't be a thing. I added a safety in the automatic action proc, but there's also a stack trace since it really should never be null.

Vi3trice commented:
I just went and replaced all path.lens.

Floorbots also had the same issue as Cleanbots where their tile routine was placed too far down and would just repath once reaching their target.

AffectedArc07 commented on code/modules/antagonists/changeling/powers/tiny_prick.dm:
```diff
@@ -41,7 +41,7 @@
 
 /datum/action/changeling/sting/can_sting(mob/user, mob/target)
-	if(!..() || !iscarbon(target) || !isturf(user.loc) || !AStar(user, target.loc, /turf/proc/Distance, cling.sting_range, simulated_only = 0))
+	if(!..() || !iscarbon(target) || !isturf(user.loc) || !get_path_to(user, target.loc, /turf/proc/Distance, cling.sting_range, simulated_only = 0))
```
This use of `/turf/proc/Distance` causes a runtime.

![image](https://user-images.githubusercontent.com/25063394/192164474-f045c97f-73c8-4c8c-9cbb-da0073ee6ecc.png)

AffectedArc07 replied:
Literally in the middle of my review this gets updated

S34NW commented:
A lot of issues showing up with bots pathing strangely.
![image](https://user-images.githubusercontent.com/12197162/192542512-e9c9edad-4d46-4e78-96e5-9085921d243d.png)
This mule was stuck in this door, when sent back and tried again it got stuck here until the door opened
![image](https://user-images.githubusercontent.com/12197162/192542801-b0fba347-fc05-483f-88b3-dd925b8ee087.png)

![image](https://user-images.githubusercontent.com/12197162/192542968-a88968b6-9e37-4a4c-853a-1c9645533cb6.png)
Bots trying to path through dense fences

![image](https://user-images.githubusercontent.com/12197162/192543080-93cea48f-8ad2-4e3c-8660-f497e4d621eb.png)
there isn't even a bot here, but the path remains

bots also teleport when moving diagonally, instead of gliding properly. It's really jarring, the diagonal paths are also quite jarring.

S34NW commented:
![image](https://user-images.githubusercontent.com/12197162/192543924-0d9b7be3-229f-4106-a927-959ddb56e496.png)
this mule was stuck "Waiting for clear path", his directional indicator facing backwards?

Mule path hud icons seem to linger even when a mule passes over a turf, unlike beepers or other bots.

![image](https://user-images.githubusercontent.com/12197162/192544576-58f23e4a-3a76-443f-9547-22f0a7f96d27.png)
bots randomly won't diagonal even though it is faster to do so here?

![image](https://user-images.githubusercontent.com/12197162/192544676-77313933-dc15-4e56-93a0-8770487953e9.png)
same here

S34NW commented:
![image](https://user-images.githubusercontent.com/12197162/192545329-c650d5d0-0067-421c-9708-521de75f65da.png)
this mule took this path

Vi3trice commented:
Oh well. You can't actually put weight on diagonals for JPS.

Gain some, lose some. Increased perf, but you gain slightly sillier bots who will do sick diagonal drifts occasionally.

S34NW commented:
There is still some minor jank for diagonal pathing, but it's mostly working well now from what I can see

lewcc commented:
I've noticed that when mulebots follow a diagonal path, they just jump to the next turf in their path rather than gliding to it. Only happens on the "new" diagonals that still look a bit scuffed.

Vi3trice commented:
Not much I can do about mules. They use `step_towards` for their movement unlike other bots, and it sort of works half the time for diagonals.

S34NW reviewed (commented):
This has seemed mostly issue-free for a while now

S34NW commented:
Issue with EDs, they get stuck trying to move onto the same tile at the same time, and just break. Beepers reroutes around them fine but they get stuck.

![image](https://user-images.githubusercontent.com/12197162/197572357-749d84b8-afdd-458b-9596-cb5445eb7969.png)

AffectedArc07 commented on code/__HELPERS/heap.dm:
```diff
@@ -11,6 +11,12 @@
 	L = new()
 	cmp = compare
 
+/datum/heap/Destroy(force, ...)
+	for(var/i in L) // because this is before the list helpers are loaded
+		qdel(i)
+	L = null
+	return ..()
```
Whats the reason for qdeleting everything in the list?

I worry this will have really bad side effects if we use heaps to do other logic with atom sorting in the future.

Vi3trice replied:
Nixing that from the port. TG did kinda pingeonhole themselves in how heaps can be used.

Still your responsibility to clear out the list in the future.

AffectedArc07 commented on code/__HELPERS/path.dm:
```diff
@@ -0,0 +1,436 @@
+ * * skip_first: Whether or not to delete the first item in the path. This would be done because the first item is the starting tile, which can break movement for some creatures.
+ * * diagonal_safety: ensures diagonal moves won't use invalid midstep turfs by splitting them into two orthogonal moves if necessary
+ */
+/proc/get_path_to(caller, end, max_distance = 30, mintargetdist, id = null, simulated_only = TRUE, turf/exclude, skip_first = TRUE, diagonal_safety = TRUE)
+	if(!caller || !get_turf(end))
+		return
+
+	var/l = SSpathfinder.mobs.getfree(caller)
+	while(!l)
+		stoplag(3)
+		l = SSpathfinder.mobs.getfree(caller)
```
A lot of `get_path_to()` calls are syncronous, and expect this to not sleep. Are you 100% sure that nothing will attempt to call this multiple times? Say a medbot wants to path somewhere, and the server is lagging like hell, what happens if SSnpc fires twice on the same bot?

Vi3trice replied:
It's hard to say honestly.

`get_path_to()` only hands out a path as a list to whatever calls it, and in the case of bots sets it as `path` and from there processes it. So if it suddently gets empty, then it'll just start over. If it's the same path but slightly behind, then either it'll correct itself (e.g. `loc == step` Alright I'm already there, next step), or if the disparity is too large, then it'll buildup `frustration` from not stepping properly and recalculate.

AffectedArc07 commented on code/controllers/subsystem/pathfinder.dm:
```diff
@@ -0,0 +1,46 @@
+SUBSYSTEM_DEF(pathfinder)
+	name = "Pathfinder"
+	init_order = INIT_ORDER_PATH
+	flags = SS_NO_FIRE
+	var/datum/flowcache/mobs
+	var/static/space_type_cache
+
+/datum/controller/subsystem/pathfinder/Initialize()
+	space_type_cache = typecacheof(/turf/space)
+	mobs = new(10)
+	return ..()
```
This being an SS not a global datum kinda pains me given it does no work and isn't init order critical.

However, thats not going to block this PR and is merely just a comment. I firmly believe any SS that doesnt require critical load order and doesnt fire should just be a misc global datum because god damn we have too many subsystems.

Take the cameranet manager for an example. Thats its own thing for good reason. SSradio however needs to die. 

---

Again, not review blocking, just me having wishful thoughts.

AffectedArc07 commented on code/game/atoms.dm:
```diff
@@ -1262,3 +1265,23 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
+ * * caller- The movable we're checking pass flags for, if we're making any such checks
+ * * no_id: When true, doors with public access will count as impassible
+ **/
+/atom/proc/CanAStarPass(obj/item/card/id/ID, to_dir, atom/movable/caller, no_id = FALSE)
+	if(caller && (caller.pass_flags & pass_flags_self))
+		return TRUE
+	. = !density
```
![image](https://user-images.githubusercontent.com/25063394/198698957-94dada84-ebc2-46e4-9eb1-f0b28f6a7518.png)

Given this only has 27 uses in the codebase (including comments), I would rather it be renamed `CanPathfindPass` or `CanPathfindThrough`, given that it isnt AStar anymore.

AffectedArc07 commented on code/modules/mob/living/simple_animal/bot/floorbot.dm:
```diff
@@ -251,6 +234,22 @@
+				target = null
+				mode = BOT_IDLE
+				return
+		else if( !bot_move(target) )
```
```suggestion
		else if(!bot_move(target))
```
Styling nit

AffectedArc07 reviewed (approved):
Ship it
